notify/jira/jira.go

```diff
@@ -0,0 +1,320 @@
+	}, nil
+}
+
+// Notify implements the Notifier interface.
+func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	key, err := notify.ExtractGroupKey(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	level.Debug(n.logger).Log("alert", key)
+
+	var (
+		tmplTextErr error
+
+		alerts       = types.Alerts(as...)
+		data         = notify.GetTemplateData(ctx, n.tmpl, as, n.logger)
+		tmplText     = notify.TmplText(n.tmpl, data, &tmplTextErr)
+		tmplTextFunc = func(tmpl string) (string, error) {
+			result := tmplText(tmpl)
+			return result, tmplTextErr
+		}
+	)
+
+	existingIssue, shouldRetry, err := n.searchExistingIssue(key, alerts.Status())
+	if err != nil {
+		return shouldRetry, fmt.Errorf("error searching existing issues: %w", err)
+	}
+
+	// Do not create new issues for resolved alerts
+	if existingIssue == nil && alerts.Status() == model.AlertResolved {
+		return false, nil
+	}
+
+	path := "issue"
+	method := http.MethodPost
+
+	if existingIssue == nil {
+		level.Debug(n.logger).Log("msg", "create new issue", "alert", key.String())
+	} else {
+		level.Debug(n.logger).Log("msg", "updating existing issue", "key", existingIssue.Key, "alert", key.String())
+
+		path += "/" + existingIssue.Key
+		method = http.MethodPut
+	}
+
+	requestBody, err := n.prepareIssueRequestBody(tmplTextFunc)
+	if err != nil {
+		return false, err
+	}
+
+	requestBody.Fields.Labels = append(requestBody.Fields.Labels, fmt.Sprintf("ALERT{%s}", key.Hash()))
+
+	for _, labelKey := range n.conf.GroupLabels {
+		if val, ok := data.GroupLabels[labelKey]; ok {
+			requestBody.Fields.Labels = append(requestBody.Fields.Labels, val)
+		}
+	}
+
+	_, shouldRetry, err = n.doAPIRequest(method, path, requestBody)
+	if err != nil {
+		return shouldRetry, fmt.Errorf("error create/update existing issues: %w", err)
+	}
+
+	if existingIssue != nil && existingIssue.Key != "" && existingIssue.Fields != nil && existingIssue.Fields.Status != nil {
+		if n.conf.ResolveTransition != "" && alerts.Status() == model.AlertResolved && existingIssue.Fields.Status.StatusCategory.Key != "done" {
+			return n.TransitionIssue(key, existingIssue.Key, n.conf.ResolveTransition)
+		} else if n.conf.ReopenTransition != "" && alerts.Status() == model.AlertFiring && existingIssue.Fields.Status.StatusCategory.Key == "done" {
+			return n.TransitionIssue(key, existingIssue.Key, n.conf.ReopenTransition)
+		}
+	}
+
+	return false, nil
+}
+
+func (n *Notifier) prepareIssueRequestBody(tmplTextFunc templateFunc) (issue, error) {
+	summary, err := tmplTextFunc(n.conf.Summary)
+	if err != nil {
+		return issue{}, fmt.Errorf("template error: %w", err)
+	}
+
+	// Recursively convert any maps to map[string]interface{}, filtering out all non-string keys, so the json encoder
+	// doesn't blow up when marshaling JIRA requests.
+	fieldsWithStringKeys, err := tcontainer.ConvertToMarshalMap(n.conf.Fields, func(v string) string { return v })
```

Comment:
I'm having a bit of trouble understanding the need for this. [`ConvertToMarshalMap`](https://pkg.go.dev/github.com/trivago/tgo@v1.0.7/tcontainer#ConvertToMarshalMap) is being passed a `map[string]any` and returns a [`MarshalMap`](https://pkg.go.dev/github.com/trivago/tgo@v1.0.7/tcontainer#MarshalMap), which is a type alias for `map[string]interface{}`, aka `map[string]any`. Am I missing something?

If this `ConvertToMarshalMap` function is _really_ needed, it seems like quite an overkill to bring in a whole new (and maybe abandoned?) third-party package, just to call a [~30-line function](https://github.com/trivago/tgo/blob/v1.0.7/tcontainer/marshalmap.go#L52). Can't that function just be "vendored" (with copyright attribution)? It's essentially just a recursive function which loops over map values and does a bit of `reflect` introspection - not exactly rocket science.

Reply:
@dswarbrick I agree with you. I was to blinded from the current jiralert. If you ask me, I would use plain strings here. I may propose an PR with would change that behavior.

Reply:
I'm a bit confused, because `n.conf.Fields` is defined as type `map[string]any`, and this is presumably populated by the config parser, reading YAML (or JSON?) config file. So if there is something that `encoding/json` does not like, presumably it's going to throw an error at _that_ stage.

I'm not too sure where the nervousness about filtering out non-string keys comes from either, since Go handles numeric keys fine:
```go
	foo := map[int]string{123: "abc", 456: "def"}
	b, _ := json.Marshal(foo)
	fmt.Println(string(b))
```
```
{"123":"abc","456":"def"}
```
Go map keys can be any type that is "comparable" - "in short, comparable types are boolean, numeric, string, pointer, channel, and interface types, and structs or arrays that contain only those types." ([ref](https://go.dev/blog/maps#key-types)). I haven't exhaustively tested all those types with JSON marshaling, but one also has to ask, how would such "exotic" types get into `n.conf.Fields` in the first place?

Reply:
nested maps will be parsed as map[any]any which is not supported in stdlib.

Reply:
the Alertmanager configuration is YAML. `yaml.Unmarshal` from `gopkg.in/yaml.v2` will be used to unmarshal to config. 

https://go.dev/play/p/yGRo4OgqXrY

Reply:
Right, I was beginning to suspect that the YAML unmarshal behaved a little differently to JSON unmarshal.

Interestingly, bumping the `yaml.v2` to `yaml.v3` would make things a lot simpler:
```
yaml.v2 unmarshal: map[string]interface {}{"field":[]interface {}{map[interface {}]interface {}{"value":0}}}
yaml.v3 unmarshal: map[string]interface {}{"field":[]interface {}{map[string]interface {}{"value":0}}}
```